core/src/SYCL/Kokkos_SYCL_Parallel_Reduce.hpp

```diff
@@ -118,116 +117,170 @@ class ParallelReduce<FunctorType, Kokkos::RangePolicy<Traits...>, ReducerType,
+      return this->get().final(value);
     }
   };
 
-  template <typename PolicyType, typename Functor>
-  void sycl_direct_launch(const PolicyType& policy,
-                          const Functor& functor) const {
+  template <typename PolicyType, typename Functor, typename Reducer>
+  void sycl_direct_launch(const PolicyType& policy, const Functor& functor,
+                          const Reducer& reducer) const {
     static_assert(ReduceFunctorHasInit<Functor>::value ==
                   ReduceFunctorHasInit<FunctorType>::value);
     static_assert(ReduceFunctorHasFinal<Functor>::value ==
                   ReduceFunctorHasFinal<FunctorType>::value);
-    static_assert(ReduceFunctorHasJoin<Functor>::value ==
-                  ReduceFunctorHasJoin<FunctorType>::value);
+    static_assert(HasJoin<Functor>::value == HasJoin<FunctorType>::value);
+    if constexpr (!std::is_same<Reducer, InvalidType>::value)
+      static_assert(HasJoin<Reducer>::value == HasJoin<ReducerType>::value);
+
+    using ReducerConditional =
+        Kokkos::Impl::if_c<std::is_same<InvalidType, Reducer>::value, Functor,
+                           Reducer>;
+    using ReducerTypeFwd = typename ReducerConditional::type;
+    using WorkTagFwd =
+        std::conditional_t<std::is_same<InvalidType, Reducer>::value, WorkTag,
+                           void>;
+    using ValueInit =
+        Kokkos::Impl::FunctorValueInit<ReducerTypeFwd, WorkTagFwd>;
+    using ValueJoin =
+        Kokkos::Impl::FunctorValueJoin<ReducerTypeFwd, WorkTagFwd>;
+    using ValueOps = Kokkos::Impl::FunctorValueOps<Functor, WorkTag>;
+
+    auto selected_reducer = ReducerConditional::select(functor, reducer);
 
     // Convenience references
     const Kokkos::Experimental::SYCL& space = policy.space();
     Kokkos::Experimental::Impl::SYCLInternal& instance =
         *space.impl_internal_space_instance();
     sycl::queue& q = *instance.m_queue;
 
-    auto result_ptr = static_cast<pointer_type>(
-        sycl::malloc(sizeof(*m_result_ptr), q, sycl::usm::alloc::shared));
-
-    value_type identity{};
-    if constexpr (!std::is_same<ReducerType, InvalidType>::value)
-      m_reducer.init(identity);
-
-    *result_ptr = identity;
-    if constexpr (ReduceFunctorHasInit<Functor>::value)
-      ValueInit::init(functor, result_ptr);
-
-    if (policy.begin() != policy.end()) {
+    // FIXME_SYCL optimize
+    constexpr size_t wgroup_size = 32;
+    std::size_t size             = policy.end() - policy.begin();
+    const auto init_size =
+        std::max<std::size_t>((size + wgroup_size - 1) / wgroup_size, 1);
+    const auto value_count =
+        FunctorValueTraits<ReducerTypeFwd, WorkTagFwd>::value_count(
+            selected_reducer);
+    const auto results_ptr =
+        static_cast<pointer_type>(Experimental::SYCLSharedUSMSpace().allocate(
+            "SYCL parallel_reduce result storage",
+            sizeof(*m_result_ptr) * std::max(value_count, 1u) * init_size));
```

Comment:
Can `value_count` actually be zero?

Reply:
Yes, we check that in https://github.com/kokkos/kokkos/blob/8c241078d0b844633cb1eae186bbb6a8a810ff5d/core/unit_test/TestReduce.hpp#L441-L461.